subsys/net/lib/lwm2m/lwm2m_rd_client.c

```diff
@@ -1091,6 +1097,44 @@ static void lwm2m_rd_client_service(struct k_work *work)
 	k_mutex_lock(&client.mutex, K_FOREVER);
 
 	if (client.ctx) {
+		if (client.suspend_thread) {
+			int ret;
+			enum lwm2m_rd_client_event event = LWM2M_RD_CLIENT_EVENT_ENGINE_SUSPENDED;
+
+			LOG_INF("Suspend client thread");
+			if (!client.ctx->connection_suspended && client.ctx->event_cb) {
+				client.ctx->event_cb(client.ctx, event);
+			}
+			lwm2m_close_suspend(client.ctx);
+			suspended_client_state = get_sm_state();
+			client.engine_state = ENGINE_SUSPENDED;
+			client.suspend_thread = false;
+			k_mutex_unlock(&client.mutex);
+			while(get_sm_state() == ENGINE_SUSPENDED) {
+				k_sleep(K_MSEC(STATE_MACHINE_UPDATE_INTERVAL_MS / 2));
+			}
```

Comment:
I'm a bit confused again, isn't the engine thread supposed to be suspended at this point? Why is this loop here?

Reply:
I change a design so don't worry.